File: apps/worker/src/routes/insights.ts
/**
 * Insights API routes
 */

import type { Env } from "../types";

interface GenerateInsightsRequest {
  company: string;
  website: string;
  token: string;
  count: number;
}

/**
 * POST /api/insights
 *
 * Proxy streaming request to runner for insights generation.
 */
export async function handleGenerateInsights(
  request: Request,
  env: Env
): Promise<Response> {
  try {
    const body = (await request.json()) as GenerateInsightsRequest;

    // Validate required fields
    if (!body.company || !body.website || !body.token) {
      return new Response(
        JSON.stringify({ error: "Missing required fields: company, website, token" }),
        { status: 400, headers: { "Content-Type": "application/json" } }
      );
    }

    // Validate count
    const count = body.count || 10;
    if (count < 1 || count > 50) {
      return new Response(
        JSON.stringify({ error: "Count must be between 1 and 50" }),
        { status: 400, headers: { "Content-Type": "application/json" } }
      );
    }

    // Proxy to runner with streaming
    const runnerResponse = await fetch(`${env.RUNNER_URL}/api/insights`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${env.RUNNER_SECRET}`,
      },
      body: JSON.stringify({
        company: body.company,
        website: body.website,
        token: body.token,
        count: count,
      }),
    });

    // Stream the response back
    return new Response(runnerResponse.body, {
      status: runnerResponse.status,
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
      },
    });
  } catch (error) {
    console.error("Error generating insights:", error);
    return new Response(
      JSON.stringify({
        error: "Failed to generate insights",
        detail: error instanceof Error ? error.message : "Unknown error",
      }),
      { status: 500, headers: { "Content-Type": "application/json" } }
    );
  }
}

File: services/runner/routes/insights.py
"""
Insights API endpoint for standalone note generation.
"""

import json
import os
import sys
from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

# Add core module to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent))

from core.insights import generate_insights_standalone
from core import AuthError, ProductboardError

from ..middleware.auth import verify_auth

router = APIRouter()


class GenerateInsightsRequest(BaseModel):
    """Request body for generating insights."""

    company: str
    website: str
    token: str
    count: int = 10


@router.post(
    "/insights",
    tags=["insights"],
)
async def generate_insights(
    request: GenerateInsightsRequest,
    _auth: str = Depends(verify_auth),
):
    """
    Generate customer feedback notes with streaming progress.

    Returns Server-Sent Events (SSE) with progress updates and final result.
    """

    def generate_events():
        try:
            for event in generate_insights_standalone(
                token=request.token,
                company=request.company,
                website=request.website,
                count=request.count,
            ):
                if event["type"] == "progress":
                    data = {
                        "current": event["current"],
                        "total": event["total"],
                        "note": event.get("note"),
                        "company": event.get("company"),
                        "phase": event.get("phase"),
                    }
                    yield f"event: progress\ndata: {json.dumps(data)}\n\n"
                elif event["type"] == "complete":
                    data = {
                        "status": "success",
                        "created": event["created"],
                        "failed": event.get("failed", 0),
                    }
                    yield f"event: complete\ndata: {json.dumps(data)}\n\n"
                elif event["type"] == "error":
                    data = {"message": event["message"]}
                    yield f"event: error\ndata: {json.dumps(data)}\n\n"

        except AuthError as e:
            error_data = {"message": str(e)}
            yield f"event: error\ndata: {json.dumps(error_data)}\n\n"
        except ProductboardError as e:
            error_data = {"message": str(e)}
            yield f"event: error\ndata: {json.dumps(error_data)}\n\n"
        except Exception as e:
            error_data = {"message": str(e)}
            yield f"event: error\ndata: {json.dumps(error_data)}\n\n"

    return StreamingResponse(
        generate_events(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
        },
    )
